### ISSUE TITLE
Implement better size profiler

### ISSUE THREAD

**Issue by binji** [WebAssembly/wabt#724]
See [Oxidizing Source Maps with Rust and WebAssembly](https://hacks.mozilla.org/2018/01/oxidizing-source-maps-with-rust-and-webassembly/).

> We found size profiling WebAssembly was more difficult than necessary. To get useful information presented meaningfully, we were forced to write our own home-grown script to post-process wasm-objdump. The script constructs the call graph, and lets us query who the callers of some function were, helping us understand why the function was emitted in the .wasm file, even if we didn’t expect it to be. It is pretty hacky, and doesn’t expose information about inlined functions. A proper WebAssembly size profiler would have helped, and would be a benefit for anyone following in our tracks.

It shouldn't be too difficult to add this functionality to wabt.

**Comment by fitzgen** [comment]
FWIW, I wrote down my wish-list for such a tool here: https://github.com/aturon/rust-wasm/issues/20#issue-289370594

TLDR: size profiling is very similar to heap profiling, just with different graphs: call graph vs heap graph and exported functions vs GC roots.

**Comment by binji** [comment]
Thanks, that's useful information! It seems like a lot of your requirements are rust-specific (or at least require standardized debug info). But I think we can provide some of the functionality in a generic way in wabt.

**Comment by fitzgen** [comment]
I think the only thing that requires new debuginfo would be inlined code analysis. The monomorphizations can be determined by symbol name (either Rust or C++ specific symbol parsing) which should be in the existing "name" section (if it isn't inlined).

All the call graph related analyses should be doable regardless of language and debug info.

**Comment by fitzgen** [comment]
https://github.com/rustwasm/twiggy implements everything I was imagining here. I think we can close this issue :)
